Parser/Parser.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+
+namespace ChameleonSystem\UpgradeHelperBundle\Parser;
+
+
+use ChameleonSystem\UpgradeHelperBundle\DataModel\CallToContainer;
+use PhpParser\NodeTraverser;
+use PhpParser\NodeVisitor\NameResolver;
+use PhpParser\ParserFactory;
+use Symfony\Component\Console\Helper\ProgressBar;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Finder\Finder;
+
+/**
+ * Parser will parse the given source code and emit lists of `CallToContainer` for further inspection.
+ * See `Validator`
+ */
+final class Parser
+{
+
+    /**
+     * parseDirectory will parse every .php file contained in $rootDirectory, including sub directories
+     * and emit all calls to the symfony container `get` method.
+     *
+     * @param string $rootDirectoy
+     * @param OutputInterface|null $output
+     * @return array
+     */
+    public function parseDirectory(string $rootDirectoy, OutputInterface $output = null): array
+    {
+        $calls = [];
+        $finder = new Finder();
+
+        $finder
+            ->in($rootDirectoy)
+            ->files()
+            ->name('*.php');
+
+        $progressBar = null;
+        if ($output !== null) {
+            $progressBar = new ProgressBar($output, $finder->count());
+        }
+
+        foreach ($finder as $file) {
+            $parsedCalls = $this->parse($file->getContents(), $file->getPathname());
```

Comment:
This is a tiny nitpick, but especially in cases like this, where something is scanned, parsed and results are calculated based on single items, I like to use Generators, just to keep potential memory issues from huge code bases (or, an infinite loop from weird symlinks in the filesystem here) in check.

This has the added benefit, that `finder` also is a generator, so in the end you can potentially loop through everything exactly once

Reply:
hmm.. i find it kind of awkward in this case because we would not yield single calls, but varying counts, including 0. Of course this could all be managed, but i decided not to in this case.

Reply:
:ok: Agreeing with you on this - just wanted to point out code style differences